shared/models/project.rb

```diff
@@ -52,29 +60,32 @@ def builds
     end
 
     # if you're using this with the fastfile parser, you need to use `relative: false`
-    def local_fastfile_path(relative: false)
+    def local_fastfile_path(relative: false, sha: nil)
       fastfile_path = nil
-      project_path = self.repo_config.local_repo_path
 
       # First assume the fastlane directory and its file is in the root of the project
-      fastfiles = Dir[File.join(project_path, "fastlane/Fastfile")]
+      fastfiles = Dir[File.join(local_repo_path(sha: sha), "**/*", "fastlane/Fastfile")]
       # If not, it might be in a subfolder
-      fastfiles = Dir[File.join(project_path, "**/fastlane/Fastfile")] if fastfiles.count == 0
+      fastfiles = Dir[File.join(local_repo_path(sha: sha), "**/fastlane/Fastfile")] if fastfiles.count == 0
 
       if fastfiles.count > 1
         logger.error("Ugh, multiple Fastfiles found, we're gonna have to build a selection in the future")
         # for now, just take the first one
       end
 
       if fastfiles.count == 0
-        logger.error("No Fastfile found at #{project_path}/fastlane/Fastfile, or any descendants")
+        logger.error("No Fastfile found at #{local_repo_path(sha: sha)}, or any descendants")
       else
         fastfile_path = fastfiles.first
         if relative
-          fastfile_path = Pathname.new(fastfile_path).relative_path_from(Pathname.new(project_path))
+          fastfile_path = Pathname.new(fastfile_path).relative_path_from(Pathname.new(local_repo_path(sha: sha)))
         end
       end
       return fastfile_path
     end
+
+    def local_repo_path(sha: nil)
+      return File.join([File.expand_path("~/.fastlane/ci/"), self.id, self.repo_config.full_name, sha, self.repo_config.name].reject { |i| i.nil? || i.empty? })
```

Comment:
let's break this down into a few steps.

Reply:
Don't know if that's really possible, I mean, the main issue here is that the `sha` might nil and the other pieces might be empty strings (shouldn't, but I fall into safety first). Do you have any idea in mind?

Reply:
I know it's not nearly as satisfying, but I think it's easier to understand at a glance like this:
```ruby
potentially_empty_path_components = [self.id, self.repo_config.full_name, sha, self.repo_config.name]
path_components = potentially_empty_path_componentsFile.reject { |i| i.nil? || i.empty? }
local_repo_path = File.join(File.expand_path("~/.fastlane/ci/"), path_components)
```